unified_planning/io/pddl_writer.py

```diff
@@ -145,72 +182,101 @@ class PDDLWriter:
+        if self.problem_kind.has_hierarchical_typing():
             user_types_hierarchy = self.problem.user_types_hierarchy
             out.write(f' (:types\n')
             stack: List['unified_planning.model.Type'] = user_types_hierarchy[None] if None in user_types_hierarchy else []
-            out.write(f'    {" ".join(self._type_name_or_object_freshname(t) for t in stack)} - object\n')
+            out.write(f'    {" ".join(self._get_mangled_name(t) for t in stack)} - object\n')
             while stack:
                 current_type = stack.pop()
                 direct_sons: List['unified_planning.model.Type'] = user_types_hierarchy[current_type]
                 if direct_sons:
                     stack.extend(direct_sons)
-                    out.write(f'    {" ".join([self._type_name_or_object_freshname(t) for t in direct_sons])} - {self._type_name_or_object_freshname(current_type)}\n')
+                    out.write(f'    {" ".join([self._get_mangled_name(t) for t in direct_sons])} - {self._get_mangled_name(current_type)}\n')
             out.write(' )\n')
         else:
-            out.write(f' (:types {" ".join([cast(UT, t).name for t in self.problem.user_types])})\n' if len(self.problem.user_types) > 0 else '')
+            out.write(f' (:types {" ".join([self._get_mangled_name(t) for t in self.problem.user_types])})\n' if len(self.problem.user_types) > 0 else '')
+
+        # Iterate the actions to retrieve domain objects
+        for a in self.problem.actions:
+            if isinstance(a, up.model.InstantaneousAction):
+                for p in a.preconditions:
+                    self.domain_objects = self.domain_objects | obe.get(p)
+                for e in a.effects:
+                    if e.is_conditional():
+                        self.domain_objects = self.domain_objects | obe.get(e.condition)
+                    self.domain_objects = self.domain_objects | obe.get(e.fluent)
+                    self.domain_objects = self.domain_objects | obe.get(e.value)
+            elif isinstance(a, DurativeAction):
+                self.domain_objects = self.domain_objects | obe.get(a.duration.lower)
+                self.domain_objects = self.domain_objects | obe.get(a.duration.upper)
+                for interval, cl in a.conditions.items():
+                    for c in cl:
+                        self.domain_objects = self.domain_objects | obe.get(c)
+                for t, el in a.effects.items():
+                    for e in el:
+                        if e.is_conditional():
+                            self.domain_objects = self.domain_objects | obe.get(e.condition)
+                        self.domain_objects = self.domain_objects | obe.get(e.fluent)
+                        self.domain_objects = self.domain_objects | obe.get(e.value)
+        if len(self.domain_objects) > 0:
+            out.write(' (:constants')
+            for ut in self.problem.user_types:
```

Comment:
You could have avoided this `for` loop is `self.domain_objects` were a `Dict[_UserType, Set[Object]]`...